app/scripts/lib/app-start.js

```diff
@@ -243,15 +247,93 @@ function (
       this._metrics.init();
     },
 
-    initializeIframeChannel: function () {
-      if (this._isIframe()) {
-        this._iframeChannel = new IframeChannel();
-        this._iframeChannel.initialize({
-          window: this._window,
-          origin: this._relier.get('origin'),
-          metrics: this._metrics
-        });
+    _getAllowedIframeOrigins: function () {
+      if (! this._isInAnIframe()) {
+        return null;
       }
+      if (this._isFxDesktop()) {
+        // If in an iframe for sync, the origin is
+        // checked against a pre-defined origin sent
+        // from the server.
+        return this._config.iframeAllowedOrigins;
+      } else if (this._isOAuth()) {
+        // If in oauth, the relier has the allowed parent origin.
+        return [this._relier.get('origin')];
+      }
+
+      return null;
+    },
+
+    _checkIframeParentOrigin: function (iframeChannel) {
+      return iframeChannel.checkParentOrigin();
+    },
+
+    _returnIframeChannelIfOriginResponds: function (origin) {
+      var self = this;
+      var iframeChannel = new IframeChannel();
+      iframeChannel.initialize({
+        window: self._window,
+        origin: origin,
+        metrics: self._metrics
+      });
+
+      // A wrapper function is used so that the wrapper can
+      // be stubbed in for unit testing.
+      return self._checkIframeParentOrigin(iframeChannel)
+        .timeout(Constants.IFRAME_CHANNEL_TIMEOUT_MS)
+        .then(function () {
+          // channel responded, this is the one we want!
+          return iframeChannel;
+        })
+        .fail(function (err) {
+          // channel is no longer needed, tear it down.
+          iframeChannel.teardown();
+
+          if (/Timed out/.test(String(err))) {
+            // a channel timeout is A-OK, that origin is not listening.
+            // The next channel will be tried.
+            return null;
+          }
+
+          throw err;
+        });
+    },
+
+    initializeIframeChannel: function () {
+      var self = this;
+      return p().then(function () {
+        if (self._isInAnIframe()) {
+          var allowedOrigins = self._getAllowedIframeOrigins() || [];
+
+          // If there is more than one allowed origin, we aren't
```

Comment:
My hope is that for prod, we define the prod firstrun server as the first origin to query. The firstrun dev server origin would be the 2nd. 

This approach should give us time to think about a better solution, perhaps by modifying the channels to not completely crap out when a message is received from an unexpected origin.